File: src/test/java/de/shevchuk/urlshortener/unitest/ShortenerServiceTest.java
package de.shevchuk.urlshortener.unitest;

import de.shevchuk.urlshortener.entity.Url;
import de.shevchuk.urlshortener.exception.UrlNotFoundException;
import de.shevchuk.urlshortener.model.ResponseUrlDto;
import de.shevchuk.urlshortener.repository.UrlRepository;
import de.shevchuk.urlshortener.service.ShortenerService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;
import java.util.Optional;

import static de.shevchuk.urlshortener.TestConstants.LONG_URL_1;
import static de.shevchuk.urlshortener.TestConstants.LONG_URL_2;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
public class ShortenerServiceTest {

    @Autowired
    private ShortenerService shortenerService;

    @Autowired
    private UrlRepository urlRepository;

    @BeforeEach
    public void resetDb() {
        urlRepository.deleteAll();
    }

    @Test
    void testShortUrlIdCreation() {
        ResponseUrlDto url = createAndSaveUrl(LONG_URL_1);
        Optional<Url> savedUrl = urlRepository.findById(url.getId());

        assertTrue(savedUrl.isPresent());
        assertEquals(url.getId(), savedUrl.get().getId());
        assertEquals(url.getUrl(), savedUrl.get().getUrl());
    }

    @Test
    void testGetAndValidateUrlById() throws UrlNotFoundException {
        ResponseUrlDto url = createAndSaveUrl(LONG_URL_1);
        String fetchedUrl = shortenerService.getAndValidateUrl(url.getId());

        assertEquals(url.getUrl(), fetchedUrl);
    }

    @Test
    void testGetAllUrls() {
        ResponseUrlDto url1 = createAndSaveUrl(LONG_URL_1);
        ResponseUrlDto url2 = createAndSaveUrl(LONG_URL_2);

        List<ResponseUrlDto> urls = shortenerService.getAllUrls();
        assertEquals(urls.size(), 2);
        assertEquals(urls.get(0).getId(), url1.getId());
        assertEquals(urls.get(0).getUrl(), url1.getUrl());
        assertEquals(urls.get(1).getId(), url2.getId());
        assertEquals(urls.get(1).getUrl(), url2.getUrl());
    }

    @Test
    void testUrlDeletion() throws UrlNotFoundException {
        ResponseUrlDto url = createAndSaveUrl(LONG_URL_1);
        shortenerService.deleteUrl(url.getId());

        Optional<Url> savedUrl = urlRepository.findById(url.getId());
        assertFalse(savedUrl.isPresent());
    }

    @Test
    void testGetAndValidateThrowUrlNotFound() {
        assertThrows(UrlNotFoundException.class, () -> shortenerService.getAndValidateUrl("NON_EXISTENT"));
    }

    @Test
    void testDeleteThrowUrlNotFound() {
        assertThrows(UrlNotFoundException.class, () -> shortenerService.deleteUrl("NON_EXISTENT"));
    }

    private ResponseUrlDto createAndSaveUrl(String longUrl) {
        return shortenerService.createShortUrl(longUrl);
    }
}

File: src/test/java/de/shevchuk/urlshortener/unitest/ShortenerServiceTest.kt
package de.shevchuk.urlshortener.unitest

import de.shevchuk.urlshortener.TestConstants.LONG_URL_1
import de.shevchuk.urlshortener.TestConstants.LONG_URL_2
import de.shevchuk.urlshortener.entity.Url
import de.shevchuk.urlshortener.exception.UrlNotFoundException
import de.shevchuk.urlshortener.model.ResponseUrlDto
import de.shevchuk.urlshortener.repository.UrlRepository
import de.shevchuk.urlshortener.service.ShortenerService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.util.*

@SpringBootTest
class ShortenerServiceTest(@Autowired val shortenerService: ShortenerService, @Autowired val urlRepository: UrlRepository){

    @BeforeEach
    fun resetDb() {
        urlRepository.deleteAll()
    }

    @Test
    fun testShortUrlIdCreation() {
        val url: ResponseUrlDto = createAndSaveUrl(LONG_URL_1)
        val savedUrl: Optional<Url> = urlRepository.findById(url.id)
        assertTrue(savedUrl.isPresent)
        assertEquals(url.id, savedUrl.get().id)
        assertEquals(url.url, savedUrl.get().url)
    }

    @Test
    @Throws(UrlNotFoundException::class)
    fun testGetAndValidateUrlById() {
        val url: ResponseUrlDto = createAndSaveUrl(LONG_URL_1)
        val fetchedUrl: String = shortenerService.getAndValidateUrl(url.id)
        assertEquals(url.url, fetchedUrl)
    }

    @Test
    fun testGetAllUrls() {
        val url1: ResponseUrlDto = createAndSaveUrl(LONG_URL_1)
        val url2: ResponseUrlDto = createAndSaveUrl(LONG_URL_2)
        val urls: List<ResponseUrlDto> = shortenerService.allUrls
        assertEquals(urls.size, 2)
        assertTrue(urls.contains(url1))
        assertTrue(urls.contains(url2))
    }

    @Test
    @Throws(UrlNotFoundException::class)
    fun testUrlDeletion() {
        val url: ResponseUrlDto = createAndSaveUrl(LONG_URL_1)
        shortenerService.deleteUrl(url.id)
        val savedUrl: Optional<Url> = urlRepository.findById(url.id)
        Assertions.assertFalse(savedUrl.isPresent)
    }

    @Test
    fun testGetAndValidateThrowUrlNotFound() {
        Assertions.assertThrows(UrlNotFoundException::class.java, Executable {
            shortenerService.getAndValidateUrl(
                "NON_EXISTENT"
            )
        })
    }

    @Test
    fun testDeleteThrowUrlNotFound() {
        Assertions.assertThrows(UrlNotFoundException::class.java, Executable {
            shortenerService.deleteUrl(
                "NON_EXISTENT"
            )
        })
    }

    private fun createAndSaveUrl(longUrl: String): ResponseUrlDto {
        return shortenerService.createShortUrl(longUrl)
    }
}